File: unalikability-js/unalikability.js
// Unalikeability returns the unalikability measure
// for `measurements`, assuming that
// `measurements` is an array describing a
// collection of measurements of a categorical
// variable (i.e: perhaps the results of running
// a multiclass classifier on a collection
// of elements of the same class).
function unalikeability(array_of_ints) {
  var empty_counts = {}
  function get_counts(l) {
    if(l.length == 1) {
      empty_counts[l[0]] = empty_counts[l[0]] || 0.0;
      empty_counts[l[0]] = empty_counts[l[0]] + 1.0
      return Object.keys(empty_counts).map(function(key){return empty_counts[key]})
    }
    empty_counts[l[0]] = empty_counts[l[0]] || 0.0;
    empty_counts[l[0]] = empty_counts[l[0]] + 1.0 
    return get_counts(l.slice(1,l.length))
  }
  counts = get_counts(array_of_ints);
  // Return 1.0 if all elements are unique
  if(counts.length == array_of_ints.length) {
    return 1.0
  }
  return (1 - counts.map(function(e){
    return Math.pow(e/array_of_ints.length, 2)
  })
    .reduce(function(a,b) {
    return a+b }))
}

File: unalikability-go/unalikability.go
package unalikability

import "math"

// Unalikeability returns the unalikability measure
// for `measurements`, assuming that
// `measurements` is an array describing a
// collection of measurements of a categorical
// variable (i.e: perhaps the results of running
// a multiclass classifier on a collection
// of elements of the same class).
func Unalikeability(measurements []int) float64 {
	unalike := 0.0
	counts := map[int]float64{}
	l := 0.0
	for _, m := range measurements {
		counts[m] += 1.0
		l += 1.0
	}
	// If every element is unique,
	// you get an unalikeability
	// of 1.0
	if float64(len(counts)) == l {
		return 1.0
	}
	for _, count := range counts {
		unalike += math.Pow(count/l, 2)
	}
	return 1 - unalike
}
